compiler/src/dotty/tools/dotc/typer/Inliner.scala

```diff
@@ -455,13 +466,99 @@ class Inliner(call: tpd.Tree, rhs: tpd.Tree)(implicit ctx: Context) {
       }
     }
 
-    override def typedApply(tree: untpd.Apply, pt: Type)(implicit ctx: Context) =
-      tree.asInstanceOf[tpd.Tree] match {
-        case Apply(Select(InlineableArg(closure(_, fn, _)), nme.apply), args) =>
-          inlining.println(i"reducing $tree with closure $fn")
-          typed(fn.appliedToArgs(args), pt)
-        case _ =>
-          super.typedApply(tree, pt)
+    override def typedApply(tree: untpd.Apply, pt: Type)(implicit ctx: Context) = {
+
+      def betaReduce(tree: Tree) = tree match {
+        case Apply(Select(cl @ closureDef(ddef), nme.apply), args) =>
```

Comment:
If the closure is a SAM type then there is no guarantee that the implement method is called `apply`, in fact it's possible that `apply` is a completely different method that has nothing to do with `ddef` so this optimization is incorrect in general, e.g.:
```scala
trait Foo {
  def apply(x: Int): Int = 42
  def realApply(x: Int): Int
}
object Test {
  def main(args: Array[String]): Unit = {
    println(((x => x): Foo).apply(10))
  }
}
```

Reply:
Right. We need to restrict this to function closures.